Deadlock on the app db (52)

### Describe the bug

Got a deadlock on the app db, I think this was when I was deleting an DB, but I may be wrong. EDIT: it's when analyzing a card

### To Reproduce

I really don't know how to repro, got it while a script that I run to initialize instances. It seems to be happening when Metabase analyzes the cards

### Expected behavior

We shouldn't cause any deadlocks

### Logs

```
2024-12-15 21:01:41,045 INFO task.sweep-query-analysis :: Recalculating potentially stale analysis
2024-12-15 21:01:42,052 ERROR task.analyze-queries :: Error analysing and updating query for Card 37
clojure.lang.ExceptionInfo: ERROR: deadlock detected
  Detail: Process 59 waits for ShareLock on transaction 2093; blocked by process 57.
Process 57 waits for ShareLock on transaction 2091; blocked by process 59.
  Hint: See server log for query details.
  Where: while locking tuple (3,6) in relation "metabase_field"
SQL statement "SELECT 1 FROM ONLY "public"."metabase_field" x WHERE "id" OPERATOR(pg_catalog.=) $1 FOR KEY SHARE OF x" {:toucan2/context-trace [["execute SQL with class com.mchange.v2.c3p0.impl.NewProxyConnection" {:toucan2.jdbc.query/sql-args ["INSERT INTO \"query_field\" (\"card_id\", \"analysis_id\", \"schema\", \"table\", \"column\", \"table_id\", \"field_id\", \"explicit_reference\") VALUES (?, ?, NULL, ?, ?, ?, ?, TRUE), (?, ?, NULL, ?, ?, ?, ?, TRUE), (?, ?, NULL, ?, ?, ?, ?, TRUE)" 37 118 "ORDERS" "PRODUCT_ID" 2 14 37 118 "ORDERS" "CREATED_AT" 2 13 37 118 "ORDERS" "QUANTITY" 2 2]}] ["resolve connection" {:toucan2.connection/connectable org.postgresql.jdbc.PgConnection}] ["resolve connection" {:toucan2.connection/connectable nil}] {:toucan2.pipeline/rf #object[clojure.core$completing$fn__8558 0x12136d20 "clojure.core$completing$fn__8558@12136d20"]} ["with compiled query" {:toucan2.pipeline/compiled-query ["INSERT INTO \"query_field\" (\"card_id\", \"analysis_id\", \"schema\", \"table\", \"column\", \"table_id\", \"field_id\", \"explicit_reference\") VALUES (?, ?, NULL, ?, ?, ?, ?, TRUE), (?, ?, NULL, ?, ?, ?, ?, TRUE), (?, ?, NULL, ?, ?, ?, ?, TRUE)" 37 118 "ORDERS" "PRODUCT_ID" 2 14 37 118 "ORDERS" "CREATED_AT" 2 13 37 118 "ORDERS" "QUANTITY" 2 2]}] ["with built query" {:toucan2.pipeline/built-query {:insert-into [:query_field], :values ((toucan2.instance/instance :model/QueryField {:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "PRODUCT_ID", :table_id 2, :field_id 14, :explicit_reference true}) (toucan2.instance/instance :model/QueryField {:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "CREATED_AT", :table_id 2, :field_id 13, :explicit_reference true}) (toucan2.instance/instance :model/QueryField {:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "QUANTITY", :table_id 2, :field_id 2, :explicit_reference true}))}}] ["with resolved query" {:toucan2.pipeline/resolved-query {}}] ["with parsed args" {:toucan2.pipeline/query-type :toucan.query-type/insert.update-count, :toucan2.pipeline/parsed-args {:rows [{:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "PRODUCT_ID", :table_id 2, :field_id 14, :explicit_reference true} {:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "CREATED_AT", :table_id 2, :field_id 13, :explicit_reference true} {:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "QUANTITY", :table_id 2, :field_id 2, :explicit_reference true}]}}] ["with model" {:toucan2.pipeline/model :model/QueryField}] ["with unparsed args" {:toucan2.pipeline/query-type :toucan.query-type/insert.update-count, :toucan2.pipeline/unparsed-args (:model/QueryField ({:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "PRODUCT_ID", :table_id 2, :field_id 14, :explicit_reference true} {:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "CREATED_AT", :table_id 2, :field_id 13, :explicit_reference true} {:card_id 37, :analysis_id 118, :schema nil, :table "ORDERS", :column "QUANTITY", :table_id 2, :field_id 2, :explicit_reference true}))}] ["resolve connection" {:toucan2.connection/connectable metabase.db.connection.ApplicationDB}] ["resolve connection" {:toucan2.connection/connectable :default}] ["resolve connection" {:toucan2.connection/connectable nil}]]}
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2725)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2412)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:502)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:419)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:194)
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:180)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.execute(NewProxyPreparedStatement.java:67)
	at toucan2.jdbc.query$reduce_jdbc_query.invokeStatic(query.clj:40)
	at toucan2.jdbc.query$reduce_jdbc_query.invoke(query.clj:22)
	at toucan2.jdbc.pipeline$transduce_execute_with_connection_primary_method_java_sql_Connection_default_default.invokeStatic(pipeline.clj:19)
	at toucan2.jdbc.pipeline$transduce_execute_with_connection_primary_method_java_sql_Connection_default_default.invoke(pipeline.clj:9)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:15)
	at methodical.impl.combo.threaded$combine_methods_thread_last$fn__19702$combined_method_thread_last__19703.invoke(threaded.clj:64)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:65)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:216)
	at toucan2.pipeline$transduce_execute$with_connection_STAR___22784$with_transaction_STAR___22785.invoke(pipeline.clj:75)
	at toucan2.connection$bind_current_connectable_fn$fn__22465.invoke(connection.clj:104)
	at metabase.db.connection$do_with_transaction_primary_method_java_sql_Connection.invokeStatic(connection.clj:180)
	at metabase.db.connection$do_with_transaction_primary_method_java_sql_Connection.invoke(connection.clj:162)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:13)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invokeStatic(connection.clj:249)
	at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invoke(connection.clj:245)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:13)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:58)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:212)
	at toucan2.pipeline$transduce_execute$with_connection_STAR___22784.invoke(pipeline.clj:74)
	at toucan2.connection$bind_current_connectable_fn$fn__22465.invoke(connection.clj:104)
	at toucan2.connection$bind_current_connectable_fn$fn__22465.invoke(connection.clj:104)
	at toucan2.jdbc.connection$do_with_connection_primary_method_java_sql_Connection.invokeStatic(connection.clj:13)
	at toucan2.jdbc.connection$do_with_connection_primary_method_java_sql_Connection.invoke(connection.clj:11)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:210)
	at toucan2.connection$do_with_connection_primary_method_.invokeStatic(connection.clj:204)
	at toucan2.connection$do_with_connection_primary_method_.invoke(connection.clj:194)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:210)
	at toucan2.pipeline$transduce_execute.invokeStatic(pipeline.clj:74)
	at toucan2.pipeline$transduce_execute.invoke(pipeline.clj:64)
	at clojure.lang.Var.invoke(Var.java:401)
	at toucan2.pipeline$transduce_compiled_query.invokeStatic(pipeline.clj:244)
	at toucan2.pipeline$transduce_compiled_query.invoke(pipeline.clj:240)
	at toucan2.pipeline$transduce_built_query.invokeStatic(pipeline.clj:252)
	at toucan2.pipeline$transduce_built_query.invoke(pipeline.clj:246)
	at toucan2.pipeline$transduce_query_primary_method_default.invokeStatic(pipeline.clj:272)
	at toucan2.pipeline$transduce_query_primary_method_default.invoke(pipeline.clj:269)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:15)
	at methodical.impl.combo.threaded$combine_methods_thread_last$fn__19702$combined_method_thread_last__19703.invoke(threaded.clj:64)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:65)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:216)
	at toucan2.pipeline$transduce_query_STAR_.invokeStatic(pipeline.clj:278)
	at toucan2.pipeline$transduce_query_STAR_.invoke(pipeline.clj:274)
	at toucan2.pipeline$transduce_with_model.invokeStatic(pipeline.clj:293)
	at toucan2.pipeline$transduce_with_model.invoke(pipeline.clj:280)
	at toucan2.pipeline$transduce_parsed.invokeStatic(pipeline.clj:309)
	at toucan2.pipeline$transduce_parsed.invoke(pipeline.clj:295)
	at toucan2.pipeline$transduce_unparsed.invokeStatic(pipeline.clj:317)
	at toucan2.pipeline$transduce_unparsed.invoke(pipeline.clj:311)
	at toucan2.pipeline$transduce_unparsed_with_default_rf.invokeStatic(pipeline.clj:374)
	at toucan2.pipeline$transduce_unparsed_with_default_rf.invoke(pipeline.clj:368)
	at toucan2.insert$insert_BANG_.invokeStatic(insert.clj:124)
	at toucan2.insert$insert_BANG_.doInvoke(insert.clj:74)
	at clojure.lang.RestFn.invoke(RestFn.java:424)
	at metabase.query_analysis$update_query_analysis_for_card_BANG_$with_connection_STAR___75567$with_transaction_STAR___75568.invoke(query_analysis.clj:157)
	at toucan2.connection$bind_current_connectable_fn$fn__22465.invoke(connection.clj:104)
	at metabase.db.connection$do_transaction$thunk__43610.invoke(connection.clj:140)
	at metabase.db.connection$do_transaction.invokeStatic(connection.clj:152)
	at metabase.db.connection$do_transaction.invoke(connection.clj:136)
	at metabase.db.connection$do_with_transaction_primary_method_java_sql_Connection.invokeStatic(connection.clj:189)
	at metabase.db.connection$do_with_transaction_primary_method_java_sql_Connection.invoke(connection.clj:162)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:13)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invokeStatic(connection.clj:249)
	at toucan2.connection$do_with_transaction_around_method_toucan2_connection_default.invoke(connection.clj:245)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:13)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:58)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:212)
	at metabase.query_analysis$update_query_analysis_for_card_BANG_$with_connection_STAR___75567.invoke(query_analysis.clj:132)
	at toucan2.connection$bind_current_connectable_fn$fn__22465.invoke(connection.clj:104)
	at toucan2.connection$bind_current_connectable_fn$fn__22465.invoke(connection.clj:104)
	at toucan2.connection$bind_current_connectable_fn$fn__22465.invoke(connection.clj:104)
	at toucan2.jdbc.connection$do_with_connection_primary_method_javax_sql_DataSource.invokeStatic(connection.clj:18)
	at toucan2.jdbc.connection$do_with_connection_primary_method_javax_sql_DataSource.invoke(connection.clj:15)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:210)
	at metabase.db.connection$do_with_connection_primary_method_default.invokeStatic(connection.clj:132)
	at metabase.db.connection$do_with_connection_primary_method_default.invoke(connection.clj:130)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:210)
	at toucan2.connection$do_with_connection_primary_method_.invokeStatic(connection.clj:204)
	at toucan2.connection$do_with_connection_primary_method_.invoke(connection.clj:194)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invokeStatic(connection.clj:118)
	at toucan2.connection$do_with_connection_around_method_toucan2_connection_default.invoke(connection.clj:106)
	at methodical.impl.combo.common$partial_STAR_$fn__19392.invoke(common.clj:12)
	at methodical.util.FnWithMeta.invoke(util.clj:46)
	at methodical.impl.standard$invoke_multifn.invokeStatic(standard.clj:55)
	at methodical.impl.standard$invoke_multifn.invoke(standard.clj:47)
	at methodical.impl.standard.StandardMultiFn.invoke(standard.clj:210)
	at metabase.query_analysis$update_query_analysis_for_card_BANG_.invokeStatic(query_analysis.clj:132)
	at metabase.query_analysis$update_query_analysis_for_card_BANG_.invoke(query_analysis.clj:124)
	at metabase.query_analysis$analyze_BANG__STAR_.invokeStatic(query_analysis.clj:239)
	at metabase.query_analysis$analyze_BANG__STAR_.invoke(query_analysis.clj:228)
	at metabase.task.analyze_queries$analyzer_loop_STAR_$fn__117936.invoke(analyze_queries.clj:51)
	at metabase.task.analyze_queries$analyzer_loop_STAR_.invokeStatic(analyze_queries.clj:50)
	at metabase.task.analyze_queries$analyzer_loop_STAR_.invoke(analyze_queries.clj:40)
	at metabase.task.analyze_queries$analyzer_loop_BANG_.invokeStatic(analyze_queries.clj:73)
	at metabase.task.analyze_queries$analyzer_loop_BANG_.invoke(analyze_queries.clj:69)
	at metabase.task.analyze_queries$analyzer_loop_BANG_.invokeStatic(analyze_queries.clj:71)
	at metabase.task.analyze_queries$analyzer_loop_BANG_.invoke(analyze_queries.clj:69)
	at metabase.task.analyze_queries.QueryAnalyzer.execute(analyze_queries.clj:82)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: org.postgresql.util.PSQLException: ERROR: deadlock detected
  Detail: Process 59 waits for ShareLock on transaction 2093; blocked by process 57.
Process 57 waits for ShareLock on transaction 2091; blocked by process 59.
  Hint: See server log for query details.
  Where: while locking tuple (3,6) in relation "metabase_field"
SQL statement "SELECT 1 FROM ONLY "public"."metabase_field" x WHERE "id" OPERATOR(pg_catalog.=) $1 FOR KEY SHARE OF x"
	... 151 more
2024-12-15 21:01:42,076 DEBUG middleware.log :: DELETE /api/database/1 204 1111ms (149 DB calls) App DB connections: 1/7 Jetty threads: 6/50 (1 idle, 0 queued) (52 total active threads) Queries in flight: 0 (0 queued) {:metabase-user-id 1}
```

### Information about your Metabase installation

```JSON
v52.2.6
```

### Severity

P1

### Additional context

This is the lab
https://github.com/paoliniluis/metabase-serialization

--- a/src/metabase/query_analysis.clj
+++ b/src/metabase/query_analysis.clj
@@ -129,41 +129,40 @@
   [{card-id :id, query :dataset_query}]
   (let [query-type (lib/normalized-query-type query)]
     (when (enabled-type? query-type)
-      (t2/with-transaction [_conn]
-        (let [analysis-id (t2/insert-returning-pk! :model/QueryAnalysis {:card_id card-id :status "running"})
-              result      (query-references query query-type)
-              safely      (fn [f] (fn [m] (update-vals (f m) truncate-string)))
-              table->row  (safely
-                           (fn [{:keys [schema table table-id]}]
-                             {:card_id     card-id
-                              :analysis_id analysis-id
-                              :schema      schema
-                              :table       table
-                              :table_id    table-id}))
-              field->row  (safely
-                           (fn [{:keys [schema table column table-id field-id explicit-reference]}]
-                             {:card_id            card-id
-                              :analysis_id        analysis-id
-                              :schema             schema
-                              :table              table
-                              :column             column
-                              :table_id           table-id
-                              :field_id           field-id
-                              :explicit_reference explicit-reference}))]
-          (if (contains? result :error)
-            ;; TODO we should track cases where the driver is disabled or not-supported differently.
-            (t2/update! :model/QueryAnalysis analysis-id {:status "failed"})
-            (do
-              (t2/insert! :model/QueryField (map field->row (:fields result)))
-              (t2/insert! :model/QueryTable (map table->row (:tables result)))
-              (t2/update! :model/QueryAnalysis analysis-id {:status "complete"})))
+      (let [analysis-id (t2/insert-returning-pk! :model/QueryAnalysis {:card_id card-id :status "running"})
+            result      (query-references query query-type)
+            safely      (fn [f] (fn [m] (update-vals (f m) truncate-string)))
+            table->row  (safely
+                         (fn [{:keys [schema table table-id]}]
+                           {:card_id     card-id
+                            :analysis_id analysis-id
+                            :schema      schema
+                            :table       table
+                            :table_id    table-id}))
+            field->row  (safely
+                         (fn [{:keys [schema table column table-id field-id explicit-reference]}]
+                           {:card_id            card-id
+                            :analysis_id        analysis-id
+                            :schema             schema
+                            :table              table
+                            :column             column
+                            :table_id           table-id
+                            :field_id           field-id
+                            :explicit_reference explicit-reference}))]
+        (if (contains? result :error)
+          ;; TODO we should track cases where the driver is disabled or not-supported differently.
+          (t2/update! :model/QueryAnalysis analysis-id {:status "failed"})
+          (do
+            (t2/insert! :model/QueryField (map field->row (:fields result)))
+            (t2/insert! :model/QueryTable (map table->row (:tables result)))
+            (t2/update! :model/QueryAnalysis analysis-id {:status "complete"})))
 
-          (t2/delete! :model/QueryAnalysis
-                      {:where [:and
-                               [:= :card_id card-id]
-                               [:not= :id analysis-id]]})
+        (t2/delete! :model/QueryAnalysis
+                    {:where [:and
+                             [:= :card_id card-id]
+                             [:not= :id analysis-id]]})
 
-          result)))))
+        result))))
 
 (defn- replaced-inner-query-for-native-card
   "Substitute new references for certain fields and tables, based upon the given mappings."

